### PR TITLE
Tickets/DM-37268: Pass filterType to ts_wep config

### PR DESCRIPTION
**Pull request by suberlak**
Updated `writeWepConfiguration` and `generateRefCatalog`, so that they use the `filterType` information, and read the magnitude limits for donut selector from `magLimits.yaml`. Added two functions to help with reading magnitude limits and converting `ref` or `""` to `g` filter. Added tests for new functions, as well as `writeWepConfiguration`.

### CODE REVIEW

**Comment by suberlak** [comment]
I ran the `black  22.10.0` on `CloseLoopTask`, because when I ran `pytest` on the ticket branch, I was getting black errors.  Yet now I see that it is `black` that fails on CI https://tssw-ci.lsst.org/job/LSST_Telescope-and-Site/job/ts_phosim/job/tickets%252FDM-37268/2/consoleFull. The CI has ` black-0.3.12`  I wonder if the two `black` versions are divergent?

**Comment by teweitsai** [comment]
My `black` version in the docker container (`lsstts/develop-env:develop`) is:

```text
[saluser@1bde3b00e858 ts_m2gui]$ black --version
black, 22.10.0 (compiled: no)
Python (CPython) 3.10.8
```

I thought the CI task will base on the docker image. Is it possible for you to reproduce this locally?

There was one huge update in Jenkins plugins and versions in yesterday. Maybe you could contact Rob if you could not reproduce this locally?

For the package I am working on recently, I have this in the CI job as well (`black-0.3.12`):
https://tssw-ci.lsst.org/job/LSST_Telescope-and-Site/job/ts_m2gui/job/tickets%252FDM-37262/2/consoleFull

```
platform linux -- Python 3.10.8, pytest-7.2.0, pluggy-1.0.0
PySide2 5.13.2 -- Qt runtime 5.12.9 -- Qt compiled 5.12.9
rootdir: /home/jenkins/workspace/d-Site_ts_m2gui_tickets_DM-37262, configfile: pyproject.toml
plugins: asdf-2.10.1, hypothesis-6.61.0, arraydiff-0.3, astropy-0.10.0, astropy-header-0.1.2, cov-4.0.0, doctestplus-0.12.1, filter-subpackage-0.1.1, flake8-1.1.1, mock-3.10.0, openfiles-0.5.0, remotedata-0.4.0, session2file-0.1.11, subtests-0.9.0, xdist-3.1.0, qt-4.1.0, xvfb-2.0.0, anyio-3.6.2, asyncio-0.18.3, black-0.3.12, mypy-0.8.0, tornasync-0.6.0.post2
asyncio: mode=strict
collected 219 items
```

Or is it possible for you to tune the format in the code to make it pass in `black 22.10.0` and `black-0.3.12`? I think there is some part of code format that `black` does not have the restrict requirement.

Thanks!

File: python/lsst/ts/phosim/CloseLoopTask.py [hunk 1]
```diff
@@ -444,6 +445,49 @@ def getFilterType(self, filterTypeName):
...
+        """Map the reference filter to the G filter.
+        Parameters
+        ----------
+        filterTypeName : str
+            Filter type name: ref (or ''), u, g, r, i, z, or y.
+        Returns
+        -------
+        filterTypeName : str
+            Mapped filter type.
+        """
```
Comment (teweitsai) [inline python/lsst/ts/phosim/CloseLoopTask.py]: Please add the new line among the fields (lines 450 and 454):

```
        """Map the reference filter to the G filter.

         Parameters
         ----------
         filterTypeName : str
             Filter type name: ref (or ''), u, g, r, i, z, or y.

         Returns
         -------
         filterTypeName : str
             Mapped filter type.
         """
```

File: python/lsst/ts/phosim/CloseLoopTask.py [hunk 2]
```diff
@@ -444,6 +445,49 @@ def getFilterType(self, filterTypeName):
...
+
+        if filterTypeName in ['ref', '']:
+            return 'g'
```
Comment (teweitsai) [inline python/lsst/ts/phosim/CloseLoopTask.py]: I thought the `black` will format it like this:

```
if filterTypeName in ("ref", ""):
    return "g"
```

Or you could do:

```
return "g" if filterTypeName in ("ref", "") else filterTypeName
```

File: python/lsst/ts/phosim/CloseLoopTask.py [hunk 3]
```diff
@@ -444,6 +445,49 @@ def getFilterType(self, filterTypeName):
...
+        else:
+            return filterTypeName
+
+    def getMagLimits(self, filterTypeName):
+        """ Read magnitude limits from the settings file.
+
+        Parameters
+        ----------
+        filterTypeName: str
```
Comment (teweitsai) [inline python/lsst/ts/phosim/CloseLoopTask.py]: `filterTypeName: str` --> `filterTypeName : str`

File: python/lsst/ts/phosim/CloseLoopTask.py [hunk 4]
```diff
@@ -444,6 +445,49 @@ def getFilterType(self, filterTypeName):
...
+            Filter type name: u, g, r, i, z, or y.
+
+        Returns
+        -------
+        magLimits: dict
```
Comment (teweitsai) [inline python/lsst/ts/phosim/CloseLoopTask.py]: `magLimits: dict` --> `magLimits : dict`

File: python/lsst/ts/phosim/CloseLoopTask.py [hunk 5]
```diff
@@ -444,6 +445,49 @@ def getFilterType(self, filterTypeName):
...
+            Dictionary of magnitude limits,
+            with keys {"low":mag_min, "high":mag_max}.
```
Comment (teweitsai) [inline python/lsst/ts/phosim/CloseLoopTask.py]: `{"low": mag_min, "high": mag_max}`

File: python/lsst/ts/phosim/CloseLoopTask.py [hunk 6]
```diff
@@ -444,6 +445,49 @@ def getFilterType(self, filterTypeName):
...
+            # Obtain filter magnitude limits
+            configDir = getConfigDir()
+            settingFilePath = os.path.join(configDir, "task/magLimitStar.yaml")
+            magLimitSettingFile = ParamReader(filePath=settingFilePath)
+
+            # The yaml file has limits for UGRIZY
+            magLimits = magLimitSettingFile.getSetting(f"filter{filterTypeName.upper()}")
+            return magLimits
+        else:
+            raise ValueError(f"This filter type ({filterTypeName}) is not supported.")
```
Comment (teweitsai) [inline python/lsst/ts/phosim/CloseLoopTask.py:494]: Please help to add the description of `ValueError` in function doc string.

File: python/lsst/ts/phosim/CloseLoopTask.py [hunk 7]
```diff
@@ -1093,21 +1153,30 @@ def runPistonWep(
...
 
-    def writeWepConfiguration(self, instName, pipelineYamlPath):
+    def writeWepConfiguration(self, instName, pipelineYamlPath,
+                              filterTypeName):
         """Write wavefront estimation pipeline task configuration.
 
         Parameters
         ----------
-        instName: `str`
+        instName: str
```
Comment (teweitsai) [inline python/lsst/ts/phosim/CloseLoopTask.py]: `instName: str` --> `instName : str`

File: python/lsst/ts/phosim/CloseLoopTask.py [hunk 8]
```diff
@@ -1093,21 +1153,30 @@ def runPistonWep(
...
             Name of the instrument this configuration is intended for.
-        pipelineYamlPath: `str`
+        pipelineYamlPath: str
```
Comment (teweitsai) [inline python/lsst/ts/phosim/CloseLoopTask.py]: `pipelineYamlPath: str` --> `pipelineYamlPath : str`

File: tests/test_closeLoopTask.py [hunk 9]
```diff
@@ -334,6 +355,69 @@ def testUseCcdImg(self):
...
+
+        # test that the reference filter
+        # gets mapped to g
+        for filterTypeName in ["ref", ""]:
+            mappedFilterName = self.closeLoopTask.mapFilterRefToG(filterTypeName)
+            self.assertEqual(mappedFilterName, "g")
+
+        # test that all other filters are
+        # mapped to themselves
+        for filterTypeName in ["ugrizy"]:
```
Comment (teweitsai) [inline tests/test_closeLoopTask.py]:
```
for filterTypeName in "ugrizy":
```

**Review by teweitsai — APPROVED** [review]
This update looks good. I noticed some thing here:

1. Sometimes you use the single quote for string and sometimes you use the double quote. I guess this is because Python does not have the strict distinction between them. But it would be better that you could unify your coding style. For other languages, there is the difference: https://www.tutorialspoint.com/single-quotes-vs-double-quotes-in-c-or-cplusplus

2. For the following codes, there is some logic difference:

```
if foo in ("a", "b", "c"):
```

or

```
if foo in ["a", "b", "c"]:
```

or 

```
if foo in "abc":
```

All of them can work in some cases (consider the support of `iteration`). The first two are `tuple` and `list`. I think the `list` is a more complex container compared with `tuple`. Therefore, I would prefer to use the `tuple` here because it is more simple:
https://en.wikipedia.org/wiki/Principle_of_least_privilege

For the final one, I think it can work just because of the property of string in Python. That means it is a immutable list or not. For example,

```
In [12]: test = "ab"

In [13]: test[0]
Out[13]: 'a'

In [14]: test[1]
Out[14]: 'b'

In [15]: test[1] = 'c'
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In [15], line 1
----> 1 test[1] = 'c'

TypeError: 'str' object does not support item assignment
```

**Comment by suberlak** [comment]
I ran `black` using version  `22.10` from `pip show black`  and `pytest --black` with  `pytest-black` version `0.3.12` from `pip show pytest-black` , to ensure compatibility with the docker CI.

**Pull request merged**